Copy-On-Write for Key and KeySet (#4691)
ElektraInitiative/libelektra: doc/contrib/mmapstorage.md

> # mmapstorage

Reviewer: Isn't here also information about the cache? Maybe call the file simply `mmap`.

Reviewer 2: IMO the cache is different thing that needs separate docs. The current cache `cache` uses `mmapstorage`, but you can use `mmapstorage` standalone and that has all the same intricacies explained here.

The `mmapstorage` plugin is a high performance storage plugin that supports full Elektra semantics.

The most important constraint for `mmapstorage` is that any structure (or bytes) that is an allocation unit (e.g. we `malloc()` the bytes needed for `KeySet` struct, so this is an unit) needs to have a flag to determine whether those bytes are actually `malloc()`ed or they are `mmap()`ed.

> The `mmapstorage` plugin only calls `munmap` in some error cases, so basically `munmap` is almost never done and the keyset is never invalidated.

Reviewer: ```suggestion
The `mmapstorage` plugin only calls `munmap` in some error cases, so basically `munmap` is never done and the keyset is never invalidated.
```

Reviewer 2: ```suggestion
The `mmapstorage` plugin only calls `munmap` in some error cases, so basically the returned keyset is never invalidated and `munmap` is never called for its data.
```

If you want it to be correct...

> During `kdbSet` the storage plugins always write to a temp file, due to how the resolver works.

Reviewer: Is this information related to the cache or the mmap storage? @mpranj

We also don't need to `mmap` the temp file here: when doing `kdbSet` we already have the `KeySet` at hand, `mmap`-ing it is not needed at all, because we have the data.
We just want to update the cache file.
The `mmap`/`munmap` in `kdbSet` are just so we can write the `KeySet` to a file in our format.
(`mmap()` is just simpler, but we could also `malloc()` a region and then `fwrite()` the stuff)

Therefore the only case where we return a `mmap()`ed KeySet should be in `kdbGet`.

When the `mmapstorage` was designed/implemented, not all structures had refcounters, so there was no way to know when a `munmap` is safe.

> This was simply out of scope at that point in time.

Reviewer: Actually we could now look at the ref counter of the KeySet, can't we? @mpranj

Reviewer 2: Doesn't really help, unless there is a callback into `mmapstorage` when `ksDel` is called or the refcount changes. `mmapstorage` could check the refcount. But it would only know whether at this moment `munmap` can be called. If it can't `munmap`, nothing changes, the data is still effectively leaked.

Also this would require that callers of `kdbGet` do a `ksIncRef`. Because right now, unless you use a binding the refcount of the `KeySet*` will very likely be 0.

Reviewer 2: However, in the refcount decision there are some ideas to make this work better.

Reviewer 3: Yes, with the refcounter we could know when munmap is safe, but we would need to iterate over all keys and data which is/was in the region.
E.g. some random key could still be used elsewhere possibly, in a different keyset, couldn't it?

Also yes, we would need a callback so we can munmap. We also need the original mmap region pointer to munmap it.

Also, the global keyset may be in that region, so again we would have to check if it is safe to munmap for all parts. So the solution won't be only a few lines of code I suspect.
